Get kubelet log from journald when kubelet is running as a systemd service in e2e tests

yifan-gu opened the pull request:
When the kubelet is running as a systemd service, we need to get the kubelet log from journald instead of /var/log/kubelet.log

I tested by running some jenkins e2e jobs against my own [coreos/e2e branch](https://github.com/yifan-gu/kubernetes/tree/coreos_gce_e2e). This works.The kubelet logs can be found [here](https://console.developers.google.com/storage/browser/kubernetes-coreos-jenkin/pr-log/kubernetes-e2e-gce-autoscaling/68/artifacts/) (kube-proxy log is still empty, because I haven't run them as static pods on that branch yet).

cc @dchen1107 @jonboulle @ixdy @zmerlynn

ixdy commented on cluster/gce/config-default.sh:
```diff
@@ -36,6 +36,11 @@ NODE_IMAGE_PROJECT=${KUBE_GCE_NODE_PROJECT:-"${MASTER_IMAGE_PROJECT}"}
 RKT_VERSION=${KUBE_RKT_VERSION:-0.5.5}
 
+# If the distro is CoreOS, kubelet will run as a systemd service.
+if [[ "${OS_DISTRIBUTION}" == "coreos" ]]; then
```
Looking at https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/pillar/systemd.sls, I worry this is a bit too fragile.

Can we instead somehow detect at run-time in the code whether the kubelet is running under systemd or not?

yifan-gu replied:
@ixdy ACK, I will give a try tomorrow, thanks!

yifan-gu replied:
@ixdy How about ssh'ing to all the hosts, and check the return code of 'sudo systemctl status kubelet.service' ?

jonboulle replied:
could you use [`go-systemd RunningFromSystemService`](https://github.com/coreos/go-systemd/blob/master/util/util.go#L166)?

jonboulle replied:
oops, sorry,missed exactly where you're checking this from - I doubt you want to expose this from within the kubelet.

yifan-gu committed:
e2e: Add support for getting kubelet logs from journald.

If there is any successful detection of kubelet.service,
fetch the kubelet logs using journalctl.

yifan-gu commented:
Pushed a new version, which checks the return code of `systemd status kubelet.service` at runtime to determine whether kubelet is running as a systemd service.
Verified by running e2e tests against gce/coreos on my branch. It successfully gets the [kubelet logs](https://00e9e64bace5df31e1648a285fffe73edbe0256aa269a43b5e-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-coreos-jenkin/o/pr-log%2Fkubernetes-e2e-gce-examples%2F70%2Fartifacts%2F104.196.47.177:22-kubelet.log?qk=AD5uMEtnLuIE4dfllGDAeFJMpQnjv5o0wQlQCOK35RipyCD_f5DAl5PcPJxJV2sHpxabdUo4Hv1xv4DUvax5S-cqE0El_hTWIcS09QrgXQY6kmVvoqhpnGeIxrKXBulVZ5bU9XPG3rgxDtu125tOW2qP0Zs-qx48lQwWSkDcSiO1Bn9a0ITa1M0Y6DPQbKN-rrpiYqj1_j3zQhbRKpAPoVCZqbHB4zHVhpWbX0VwkMlMJSfarXr3uK0zObnWBOxnQCM6NjZeiyeW6adzqF2qLbCHkNooTdcl9TIjnwT0uzf9HRi31oK0V6l7o1dMjCS4YZfCkWIyUEan8kYvDGjVdL2drPd7dPEqwYCNERiIrjPc2fcBPgfROupqc80dICZFxaEAEe76dhgzGhFlLKSSD9kLgAK61-yCXtfq0yYd8nBD1t38BBuLR7gCLqORW590uD6PhkJY2vRGYLM_f7lV_nPtyDzcRSOO0iBT3ZW-U-etSp2bCsrtOcBNgIl2ARHO9UhqjMDS4mFYJOSxqHoUXxcphUqHlwm2Kw_pcNppMISEXLvowQBtKo8gl05x_adefXjF7fPxlc1AKvCcjpq1obvCmRGQKYLAKb-E2KPZ3_XgIqdUV12QcUC_Eq-qX8eVMhlWFh8ir4xrXZ2EZGraIj9O_IFrGEJY4eJP1xHsvXfrES-XeIpCiqIjdTg6x29X72wkXCaf41cOj6wPZiEaIfZzRfako5GInlwm4ehrMXzkPRRLzu-am2AQ-aTGRIYuWbtw44wkhgJUzhaI7WTSsM0SsU3KmoJGDM-s0am3tllaSi8fBw2bsujErcxUdWsgUAsK0RuzNlC7qtvUNB5BE7lQfc-fvCwcVlXUQfbqLHksdzAhm3MzurPYnHfOL7wHGkGjXu6KIGtCKiv_apXbHidnrSvX06wkL6b_UJ0cAHDoQk6EPPg3o1E)

yifan-gu commented:
@ixdy This hardcodes "kubelet.service" as the name of the kubelet systemd service. Might not be the best solution, but should be better than my original one. Feedback is welcome :)

ixdy commented on test/e2e/core.go:
```diff
@@ -97,3 +123,29 @@ func logCore(cmds []command, hosts []string, dir, provider string) {
 	wg.Wait()
 }
+
+func findKubeletSystemdService(provider string, hosts ...string) (bool, error) {
```
two minor nits:
* maybe "isUsingSystemdKubelet" is a better name for this function?
* you never use the error return value - should this return only bool?

yifan-gu replied:
@ixdy What do you mean by "never use the error return value"? If there's an error, a message will be printed.

ixdy commented on test/e2e/core.go:
```diff
@@ -97,3 +123,29 @@ func logCore(cmds []command, hosts []string, dir, provider string) {
 	wg.Wait()
 }
+
+func isUsingSystemdKubelet(provider string, hosts ...string) (bool, error) {
```
there are only two return points from this function on lines 147 and 150, and neither one returns an error.

yifan-gu replied:
@ixdy Got you. Removed the returned error. Thanks!

ixdy commented on test/e2e/core.go:
```diff
@@ -50,11 +50,18 @@ func CoreDump(dir string) {
+	cmds := []command{{"cat /var/log/kube-proxy.log", "kube-proxy"}}
+
+	found := isUsingSystemdKubelet(provider, hosts...)
+	if found {
```
one more minor nit (sorry): rather than using the variable `found`, maybe just have
```
if isUsingSystemdKubelet(provider, hosts...) {
   ...
} else {
  ...
}
```

similarly for the master.

yifan-gu replied:
@ixdy done

yifan-gu commented:
@k8s-bot Test this please

derekwaynecarr commented on test/e2e/core.go:
```diff
@@ -50,11 +50,16 @@ func CoreDump(dir string) {
-		{"cat /var/log/supervisor/supervisord.log", "supervisord"},
+	cmds := []command{{"cat /var/log/kube-proxy.log", "kube-proxy"}}
+	if isUsingSystemdKubelet(provider, hosts...) {
+		cmds = append(cmds, command{"sudo journalctl --output=cat -u kubelet.service", "kubelet"})
```
can we let the name of kubelet.service get passed down somehow rather than hard-code it?

ixdy replied:
we could factor out "kubelet.service" into a constant, sure. how likely is this to change, though?

derekwaynecarr commented on test/e2e/core.go:
```diff
@@ -97,3 +110,29 @@ func logCore(cmds []command, hosts []string, dir, provider string) {
+func isUsingSystemdKubelet(provider string, hosts ...string) bool {
+	wg := &sync.WaitGroup{}
+	results := make([]int, len(hosts))
+	cmd := "sudo systemctl status kubelet.service"
```
same here, it would be ideal if the service name was not hardcoded.

yifan-gu commented:
> can we let the name of kubelet.service get passed down somehow rather than hard-code it?

I feel it's better. Also if pass the name of the service, do we still need to check kubelet service at runtime? 
For example, If we pass a flag `--kubelet-service-name=""`, can we assume if the flag is non-empty, the kubelet is running as a service?
@ixdy

ixdy commented:
I think this is probably fine as-is.

yifan-gu commented:
> I think this is probably fine as-is.
@ixdy Sorry, do you mean the current approach (which hardcodes the kubelet.service) is fine, or passing `--kubelet-service-name` flag is fine?

ixdy commented:
Approach as currently implemented (everything hard-coded) is fine with me.

jonboulle commented on test/e2e/core.go:
```diff
@@ -97,3 +110,29 @@ func logCore(cmds []command, hosts []string, dir, provider string) {
+			if err != nil {
+				fmt.Printf("Error running command: %v\n", err)
+			}
+			results[i] = result.Code
```
This is going to set to 0 if err != nil, is that right/expected?

yifan-gu replied:
@jonboulle So in this case we assume the kubelet is running as systemd service even when there are errors during SSH().

ixdy replied:
@jonboulle raises a good point. if a node has gone away (and thus this ssh call fails), we don't want that to influence the choice of logging location for the rest of the nodes.

I think we should probably be using only successful ssh connections to detect whether nodes are using systemd or not. Could probably even shortcut - return true/false for the first host that we successfully connect to, rather than checking all. I imagine that's the easiest fix.

yifan-gu replied:
@ixdy Actually SSH failures doens't affect this as the code is still 0 when there is an error. But I made a change so instead of assume "kubelet is running as systemd service", now we assume "kubelet is not running as systemd service".

yifan-gu commented:
@ixdy So is this OK to merge or you have more concerns?

yifan-gu commented:
ping @ixdy

yifan-gu commented:
@ixdy Updated. Now:
If there's any successful detects of kubelet.service, using journalctl to fetch the log.

ixdy commented:
LGTM, thanks
